python/tvm/relax/training/optimizer.py

```diff
@@ -0,0 +1,582 @@
+# Floating point error would lead to a few problems, such as making assert_structural_equal not
+# passed in unit tests
+def _high_precision_subtract(lhs, rhs):
+    return float(Decimal(str(lhs)) - Decimal(str(rhs)))
+
+
+class SGD(Optimizer):
+    """Implements stochastic gradient descent.
+
+    The returned function is equivalent to the following numpy code:
+
+    .. code-block:: python
+        def SGD(param_tuple, grad_tuple, state_tuple):
+            num_steps = state_tuple[0]
+            state_tuple[0] = num_steps + 1
+            for i in range(len(param_tuple)):
+                param = param_tuple[i]
+                grad = grad_tuple[i]
+                param_tuple[i] = param - lr * (grad + weight_decay * param)
+            return param_tuple, state_tuple
+
+    Parameters
+    ----------
+    lr : float
+        learning rate
+
+    weight_decay : Optional[float]
+        weight decay (L2 penalty) (default: 0)
+    """
+
+    def __init__(self, param_list, lr, weight_decay=0):
+        super().__init__(param_list)
+        self.lr = float(lr)
+        self.weight_decay = float(weight_decay)
+
+    @property
+    def state(self):
+        """The state of SGD is `(num_steps,)`.
+
+        Returns
+        -------
+        res : ADT
+            The state of SGD.
+        """
+        if self._state is None:
+            self._state = tuple_object(
+                (
+                    # num_steps = 0
+                    tvm.nd.array(np.zeros((), "int64")),
+                )
+            )
+        return self._state
+
+    @state.setter
+    def state(self, value):
+        self._state = value
+
+    def get_function(self) -> Function:
+        plist = self._param_list
+        len_param = len(plist)
+        dtype = self._dtype
+
+        # input variables
+        param_var = Var("params", rx.TupleStructInfo([p.struct_info for p in plist]))
+        grad_var = Var("gradients", rx.TupleStructInfo([p.struct_info for p in plist]))
+        state_var = Var("optim_states", rx.TupleStructInfo([rx.TensorStructInfo((), "int64")]))
+
+        # constants
+        lr = rx.const(self.lr, dtype)
+        weight_decay = rx.const(self.weight_decay, dtype)
+        one = rx.const(1, "int64")
+
+        builder = rx.BlockBuilder()
+        with builder.function("SGD", [param_var, grad_var, state_var]):
+            with builder.dataflow():
+                # get variables in tuples
+                param_var_list = [
+                    builder.emit(rx.TupleGetItem(param_var, i)) for i in range(len_param)
+                ]
+                grad_var_list = [
+                    builder.emit(rx.TupleGetItem(grad_var, i)) for i in range(len_param)
+                ]
+                state_var_list = [builder.emit(rx.TupleGetItem(state_var, 0))]
+
+                # computation logic
+                state_var_list[0] = builder.emit(add(state_var_list[0], one))
```

Comment:
I suggest using some other name like `new_state_var`, to distinguish the post-update state with the input state.